nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/main/java/org/apache/nifi/csv/JacksonCSVRecordReader.java

```diff
@@ -0,0 +1,257 @@
+
+        this.schema = schema;
+        this.logger = logger;
+        this.hasHeader = hasHeader;
+        this.ignoreHeader = ignoreHeader;
+        final DateFormat df = dateFormat == null ? null : DataTypeUtils.getDateFormat(dateFormat);
+        final DateFormat tf = timeFormat == null ? null : DataTypeUtils.getDateFormat(timeFormat);
+        final DateFormat tsf = timestampFormat == null ? null : DataTypeUtils.getDateFormat(timestampFormat);
+
+        LAZY_DATE_FORMAT = () -> df;
+        LAZY_TIME_FORMAT = () -> tf;
+        LAZY_TIMESTAMP_FORMAT = () -> tsf;
+
+        final Reader reader = new InputStreamReader(new BOMInputStream(in));
+
+        CsvSchema.Builder csvSchemaBuilder = CsvSchema.builder()
+                .setColumnSeparator(csvFormat.getDelimiter())
+                .setLineSeparator(csvFormat.getRecordSeparator())
+                // Can only use comments in Jackson CSV if the correct marker is set
+                .setAllowComments("#" .equals(CharUtils.toString(csvFormat.getCommentMarker())))
+                // The call to setUseHeader(false) in all code paths is due to the way Jackson does data binding/mapping. Missing or extra columns may not
+                // be handled correctly when using the header for mapping.
+                .setUseHeader(false);
+
+        csvSchemaBuilder = (csvFormat.getQuoteCharacter() == null) ? csvSchemaBuilder : csvSchemaBuilder.setQuoteChar(csvFormat.getQuoteCharacter());
+        csvSchemaBuilder = (csvFormat.getEscapeCharacter() == null) ? csvSchemaBuilder : csvSchemaBuilder.setEscapeChar(csvFormat.getEscapeCharacter());
+
+        if (hasHeader) {
+            if (ignoreHeader) {
+                csvSchemaBuilder = csvSchemaBuilder.setSkipFirstDataRow(true);
+            }
+        }
+
+        CsvSchema csvSchema = csvSchemaBuilder.build();
+
+        // Add remaining config options to the mapper
+        List<CsvParser.Feature> features = new ArrayList<>(3);
+        features.add(CsvParser.Feature.INSERT_NULLS_FOR_MISSING_COLUMNS);
+        if (csvFormat.getIgnoreEmptyLines()) {
+            features.add(CsvParser.Feature.SKIP_EMPTY_LINES);
+        }
+        if (csvFormat.getTrim()) {
+            features.add(CsvParser.Feature.TRIM_SPACES);
+        }
+
+        ObjectReader objReader = mapper.readerFor(String[].class)
+                .with(csvSchema)
+                .withFeatures(features.toArray(new CsvParser.Feature[3]));
+
+        recordStream = objReader.readValues(reader);
+    }
+
+    @Override
+    public Record nextRecord(final boolean coerceTypes, final boolean dropUnknownFields) throws IOException, MalformedRecordException {
+        final RecordSchema schema = getSchema();
+
+        if (recordStream.hasNext()) {
+            String[] csvRecord = null;
+            try {
+                csvRecord = recordStream.next();
+            } catch (Exception e) {
```

Comment:
This gives me pause. I think if we catch IOException, we should let it fly. Otherwise, we should wrap the Exception in a MalformedRecordException

Reply:
Agreed, that might be left over from a debugging exercise. Definitely not great as-is, will change.